File: benchmarks/src/data/__fixtures__/native/workload.cpp
// Reference workloads for the cross-language groups.
//
// Usage: workload <noop|queue|heap> <n>
//
// `noop` gives the harness the process startup cost to subtract. The checksum
// is printed so the caller can verify every language computed the same answer.

#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <functional>
#include <queue>
#include <vector>

namespace
{

	constexpr std::uint32_t SEED = 0x2545F491u;

	std::uint32_t xorshift32(std::uint32_t state)
	{
		state ^= state << 13;
		state ^= state >> 17;
		state ^= state << 5;
		return state;
	}

	std::uint32_t queue_workload(std::uint32_t count)
	{
		std::deque<std::uint32_t> queue;
		std::uint32_t total = 0;
		for (std::uint32_t index = 0; index < count; ++index)
		{
			queue.push_back(index);
			if (index & 1u)
			{
				total += queue.front();
				queue.pop_front();
			}
		}
		while (!queue.empty())
		{
			total += queue.front();
			queue.pop_front();
		}
		return total;
	}

	std::uint32_t heap_workload(std::uint32_t count)
	{
		std::priority_queue<std::uint32_t, std::vector<std::uint32_t>,
												std::greater<std::uint32_t>>
				heap;
		std::uint32_t state = SEED;
		for (std::uint32_t index = 0; index < count; ++index)
		{
			state = xorshift32(state);
			heap.push(state);
		}
		std::uint32_t total = 0;
		while (!heap.empty())
		{
			total += heap.top();
			heap.pop();
		}
		return total;
	}

} // namespace

int main(int argc, char **argv)
{
	if (argc < 2)
	{
		std::fprintf(stderr, "usage: workload <noop|queue|heap> <n>\n");
		return 2;
	}

	const char *mode = argv[1];
	const std::uint32_t count =
			argc > 2 ? static_cast<std::uint32_t>(std::strtoul(argv[2], nullptr, 10))
							 : 0;

	if (std::strcmp(mode, "noop") == 0)
	{
		std::printf("0\n");
	}
	else if (std::strcmp(mode, "queue") == 0)
	{
		std::printf("%u\n", queue_workload(count));
	}
	else if (std::strcmp(mode, "heap") == 0)
	{
		std::printf("%u\n", heap_workload(count));
	}
	else
	{
		std::fprintf(stderr, "unknown mode: %s\n", mode);
		return 2;
	}

	return 0;
}

File: benchmarks/src/data/__fixtures__/native/workload.rs
// Reference workloads for the cross-language groups.
//
// Usage: workload <noop|queue|heap> <n>
//
// `noop` gives the harness the process startup cost to subtract. The checksum
// is printed so the caller can verify every language computed the same answer.

use std::cmp::Reverse;
use std::collections::{BinaryHeap, VecDeque};
use std::process::ExitCode;

const SEED: u32 = 0x2545_F491;

fn xorshift32(mut state: u32) -> u32 {
	state ^= state << 13;
	state ^= state >> 17;
	state ^= state << 5;
	state
}

fn queue_workload(count: u32) -> u32 {
	let mut queue: VecDeque<u32> = VecDeque::new();
	let mut total: u32 = 0;
	for index in 0..count {
		queue.push_back(index);
		if index & 1 == 1 {
			total = total.wrapping_add(queue.pop_front().unwrap_or(0));
		}
	}
	while let Some(value) = queue.pop_front() {
		total = total.wrapping_add(value);
	}
	total
}

fn heap_workload(count: u32) -> u32 {
	let mut heap: BinaryHeap<Reverse<u32>> = BinaryHeap::new();
	let mut state = SEED;
	for _ in 0..count {
		state = xorshift32(state);
		heap.push(Reverse(state));
	}
	let mut total: u32 = 0;
	while let Some(Reverse(value)) = heap.pop() {
		total = total.wrapping_add(value);
	}
	total
}

fn main() -> ExitCode {
	let args: Vec<String> = std::env::args().collect();
	let Some(mode) = args.get(1) else {
		eprintln!("usage: workload <noop|queue|heap> <n>");
		return ExitCode::from(2);
	};
	let count: u32 = args.get(2).and_then(|value| value.parse().ok()).unwrap_or(0);

	match mode.as_str() {
		"noop" => println!("0"),
		"queue" => println!("{}", queue_workload(count)),
		"heap" => println!("{}", heap_workload(count)),
		other => {
			eprintln!("unknown mode: {other}");
			return ExitCode::from(2);
		}
	}

	ExitCode::SUCCESS
}
